### DISCUSSION TITLE
Best way to read updated value right after calling the setter?

### DISCUSSION THREAD

**Discussion by SzabKel** [pmndrs/jotai#1376] [Q&A] [answered]
I am using jotai mostly for global state. I define a state interface for a specific functionality (like a sidebar for example, a property is " is it open or closed" for example). Share it between components (multiple components can change it). 

I keep running into this issue, where I would like to use the value "returned by the setter" (it does not return anything and there is no callback with the updated value). 

- I have to keep the other property values, so I can't do:
```javascript
const newValue
set(newValue)
doOtherStuff(newValue)
```
- I also change the state from different components, so I have to do:
```javascript
set(prev => {
  return {...prev, changeValue: X}
})
```

How do I get the current value after the setter executed without waiting for a new render (and using `useEffect`)?

Full example: https://codesandbox.io/s/jotai-get-current-value-after-setter-eql2h8?file=/src/App.tsx

```javascript
import { atom, useAtom } from "jotai";
import { useRef } from "react";
import "./styles.css";

interface StateObj {
  /**
   * This number can be changed by
   * other components
   */
  num: number;
  /**
   * other properties can be changed
   * by different components as well
   */
  keepThisData: string;
}

const numAtom = atom<StateObj>({
  num: 0,
  keepThisData: "Important information"
});

export default function App() {
  const [state, setState] = useAtom(numAtom);

  const numRef = useRef(state.num);
  numRef.current = state.num;

  return (
    <div className="App">
      <button
        onClick={() => {
          setState((prev) => {
            return {
              ...prev,
              num: prev.num + 1
            };
          });

          // how to get the updated 'num' value here?
          // or how to update the numRef?

          alert("Num value during onClick: " + state.num);
          alert("NumRef value during onClick: " + numRef.current);
        }}
      >
        Action
      </button>
      <p>Num value on render: {state.num}</p>
    </div>
  );
}
```

I could do this, but it feels wrong:
```javascript
setState((prev) => {
	numRef.current = prev.num + 1;
	return {
		...prev,
		num: prev.num + 1
	};
});
```

**Comment by dai-shi** [comment, marked as answer]
It's always a little tricky if we want to communicate between atoms and others such as props, useState, and useRef.
Generally, we should create a new atom based on them.
Here's how it works: https://codesandbox.io/s/jotai-get-current-value-after-setter-forked-vw9vfj?file=/src/App.tsx

Alternatively, `useAtomCallback` works too, but I would prefer `useMemo(() => atom(...), [...])` if works.
